src/blockly/generators/propc/variables.js

```diff
@@ -165,186 +165,131 @@ Blockly.Blocks.array_get = {
+        this.buildArrayMenu(this.arrayList);
 
-        // TODO: What is this code doing?
-        if (m && m === oldVarName && newVarName) {
-            this.setFieldValue(newVarName, 'VAR');
-        } else if (m) {
-            this.setFieldValue(m, 'VAR');
+        // update the menu on the block
+        if (newValue && currentValue && currentValue === oldValue) {
+            this.setFieldValue(newValue, 'VAR');
+        } else if (currentValue) {
+            this.setFieldValue(currentValue, 'VAR');
         }
     },
+    onchange: function (event) {
+        var blockType = null;
+        var oldValue = null;
+        var newValue = null;
+        if (event.type === Blockly.Events.BLOCK_DELETE) {
+            // The block's fields can't be accessed after it's deleted, so we 
+            // have to dive into it's XML, since that's all that's preserved 
+            if (event.oldXml.attributes.type.nodeValue === 'array_init') {
+                blockType = 'array_init';
+                oldValue = event.oldXml.children.VAR.firstChild.data; // Block field previous value
+            }        
+        } else if (event.type === Blockly.Events.BLOCK_CREATE) {
+            var eventBlock = Blockly.getMainWorkspace().getBlockById(event.blockId);
+            if (eventBlock && eventBlock.type === 'array_init') {
+                blockType = 'array_init';
+                newValue = eventBlock.getFieldValue('VAR');  // Block field new value
+            }
+        } else if (event.type === Blockly.Events.BLOCK_CHANGE) {
+            var eventBlock = Blockly.getMainWorkspace().getBlockById(event.blockId);
+            if (eventBlock && eventBlock.type === 'array_init') {
+                blockType = 'array_init';
+                newValue = event.newValue;  // Block field new value
+                oldValue = event.oldValue;  // Block field previous value
+            }
+        }
+        if (blockType === 'array_init') {
+            this.updateArrayMenu(oldValue, newValue);
+        }
 
-    onchange: function () {
         var code = null;
-        var elmnts = parseInt(this.getFieldValue('NUM'), 10);
-        var allBlocks = Blockly.getMainWorkspace().getAllBlocks();
-        if (allBlocks.toString().indexOf('array initialize ' + this.getFieldValue('VAR')) > -1) {
-            var initStr = '';
-            for (var ij = 0; ij < allBlocks.length; ij++) {
-                var f_start = allBlocks[ij].toString().indexOf('array initialize ' + this.getFieldValue('VAR'));
-                if (f_start > -1) {
-                    initStr = allBlocks[ij].toString().substring(f_start).replace(/[^0-9]/g, "");
-                    break;
-                }
+        var elementCount = null;
+        if (this.type === 'array_get' || this.type === 'array_set' ) {
+            var en = Blockly.propc.valueToCode(this, 'NUM', Blockly.propc.ORDER_NONE) || '0';
```

Comment:
Is this replacing non-numeric characters with ""?  It would be helpful to add a single line comment when using regex to help the uninitiated to understand what the operation is trying to accomplish.

Reply:
Will do.